modules/tinymce/src/core/main/ts/Rtc.ts

```diff
@@ -0,0 +1,296 @@
+import { Event, Node as DomNode, Range } from '@ephox/dom-globals';
+import { Fun, Obj, Option, Type } from '@ephox/katamari';
+import Editor from './api/Editor';
+import Node from './api/html/Node';
+import Serializer from './api/html/Serializer';
+import * as FilterNode from './html/FilterNode';
+import * as Operations from './undo/Operations';
+import { Index, Locks, UndoBookmark, UndoLevel, UndoLevelType, UndoManager } from './undo/UndoManagerTypes';
+import * as ApplyFormat from './fmt/ApplyFormat';
+import * as RemoveFormat from './fmt/RemoveFormat';
+import * as ToggleFormat from './fmt/ToggleFormat';
+import { RangeLikeObject } from './selection/RangeTypes';
+import { FormatRegistry } from './fmt/FormatRegistry';
+import { Content } from './content/EditorContent';
+import { GetContentArgs, ContentFormat, getContentInternal } from './content/GetContentImpl';
+import { SetContentArgs, setContentInternal } from './content/SetContentImpl';
+import { insertHtmlAtCaret } from './content/InsertContentImpl';
+import { getSelectedContentInternal } from './selection/GetSelectionContentImpl';
+
+const isTreeNode = (content: any): content is Node => content instanceof Node;
+
+const isSupportedContentFormat = (format: string) => format !== 'text';
+
+/** API implemented by the RTC plugin */
+interface RtcRuntimeApi {
+  undo: () => void;
+  redo: () => void;
+  hasUndo: () => boolean;
+  hasRedo: () => boolean;
+  transact: (fn: () => void) => void;
+  applyFormat: (format: string, vars: Record<string, string>) => void;
+  removeFormat: (format: string, vars: Record<string, string>) => void;
+  toggleFormat: (format: string, vars: Record<string, string>) => void;
+  getContent: () => Node | null;
+  setContent: (node: Node) => void;
+  insertContent: (node: Node) => void;
+  getSelectedContent: () => Node | null;
+  getRawModel: () => any;
+  isRemote: boolean;
+}
+
+/** A copy of the TinyMCE api definitions that the plugin overrides  */
+interface RtcAdaptor {
```

Comment:
There seems to be a lot of duplication of types here. Do we actually want that, or should we be reusing the core types? eg `Formatter['apply']`

Reply:
They didn't start out as straight duplication, but they were refactored into this and you're right that's exactly what we have

- [ ] reference types instead of duplicate

Reply:
actually that _will_ give us a circular dep, because `Formatter.apply` calls this code now 🤔 what to do, what to do 🤔

Reply:
Oh fun 😂 I didn't quite get the full issue, but could we pull the types out into their own file? eg `FormatterTypes` or something?

Reply:
ah good point, if circles cause a problem we can just move the interfaces into their own file (and if there isn't an interface, make one).

Reply:
Can we even reuse types these are not full implementations of the interfaces. I guess we could pull each function out into separate types and reuse in the interfaces and in this but I think that would be a bit harder to read.

Reply:
You can, so this for example the formatter could be done via:
```
formatter: {
    applyFormat: Formatter['apply'];
    removeFormat: Formatter['remove'];
    toggleFormat: Formatter['toggle'];
  };
```

Reply:
Didn't know about that feature thanks.